cmd/thanos/remote-write-receive.go

```diff
@@ -0,0 +1,270 @@
+		Default("0.0.0.0:19291").String()
+
+	dataDir := cmd.Flag("tsdb.path", "Data directory of TSDB.").
+		Default("./data").String()
+
+	m[name] = func(g *run.Group, logger log.Logger, reg *prometheus.Registry, tracer opentracing.Tracer, _ bool) error {
+		peer, err := newPeerFn(logger, reg, false, "", false)
+		if err != nil {
+			return errors.Wrap(err, "new cluster peer")
+		}
+		return runReceiver(
+			g,
+			logger,
+			reg,
+			tracer,
+			*grpcBindAddr,
+			*cert,
+			*key,
+			*clientCA,
+			*httpMetricsBindAddr,
+			*remoteWriteAddress,
+			*dataDir,
+			peer,
+		)
+	}
+}
+
+func runReceiver(
+	g *run.Group,
+	logger log.Logger,
+	reg *prometheus.Registry,
+	tracer opentracing.Tracer,
+	grpcBindAddr string,
+	cert string,
+	key string,
+	clientCA string,
+	httpMetricsBindAddr string,
+	remoteWriteAddress string,
+	dataDir string,
+	peer cluster.Peer,
+) error {
+	level.Info(logger).Log("msg", "setting up receiver")
+
+	var metadata = &promMetadata{
+		// Start out with the full time range. The shipper will constrain it later.
+		// TODO(fabxc): minimum timestamp is never adjusted if shipping is disabled.
+		mint: 0,
+		maxt: math.MaxInt64,
+	}
+
+	tsdbCfg := &tsdb.Options{
+		Retention:        model.Duration(time.Hour * 24 * 15),
+		NoLockfile:       true,
+		MinBlockDuration: model.Duration(time.Hour * 2),
+		MaxBlockDuration: model.Duration(time.Hour * 2),
+	}
+
+	localStorage := &tsdb.ReadyStorage{}
+	receiver := receive.NewWriter(log.With(logger, "component", "receive-writer"), localStorage)
+	webHandler := receive.NewHandler(log.With(logger, "component", "remote-receive-handler"), &receive.Options{
+		Receiver:      receiver,
+		ListenAddress: remoteWriteAddress,
+		Registry:      reg,
+		ReadyStorage:  localStorage,
+	})
+
+	// Start all components while we wait for TSDB to open but only load
+	// initial config and mark ourselves as ready after it completed.
+	dbOpen := make(chan struct{})
+
+	// sync.Once is used to make sure we can close the channel at different execution stages(SIGTERM or when the config is loaded).
+	type closeOnce struct {
+		C     chan struct{}
+		once  sync.Once
+		Close func()
+	}
+	// Wait until the server is ready to handle reloading.
+	reloadReady := &closeOnce{
+		C: make(chan struct{}),
+	}
+	reloadReady.Close = func() {
+		reloadReady.once.Do(func() {
+			close(reloadReady.C)
+		})
+	}
+
+	level.Debug(logger).Log("msg", "setting up joining cluster")
+	{
+		ctx, cancel := context.WithCancel(context.Background())
+		g.Add(func() error {
+			mint, maxt := metadata.Timestamps()
+			if peer != nil {
+				if err := peer.Join(cluster.PeerTypeSource, cluster.PeerMetadata{
+					Labels:  metadata.LabelsPB(),
+					MinTime: mint,
+					MaxTime: maxt,
+				}); err != nil {
+					return errors.Wrap(err, "join cluster")
+				}
+			}
+
+			<-ctx.Done()
+			return nil
+		}, func(err error) {
+			cancel()
+			peer.Close(2 * time.Second)
+		})
+	}
```

Comment:
can we remove this whole block? Will receive need to peer via gossip? If we are removing gossip in the near future I think we should not add it to any new components and instead help people migrate to the new approach.